modules/core/shared/src/main/scala/gem/math/JulianDate.scala

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gem.math
+
+import cats.{Eq, Show}
+
+import java.time.{Instant, LocalDateTime}
+import java.time.ZoneOffset.UTC
+
+/** Astronomical time representation of continuous days since noon, November 24,
+  * 4714 BC.
+  *
+  * @param dayNumber      whole Julian Day number
+  * @param nanoAdjustment fraction of a day adjustment in nanoseconds
+  *
+  * @see The Wikipedia [[https://en.wikipedia.org/wiki/Julian_day article]]
+  */
+sealed abstract case class JulianDate(
+  dayNumber:      Int,
+  nanoAdjustment: Long
+) {
+
+  import JulianDate.{ NanoPerDay, MaxAdjustment, MinAdjustment }
+
+  // Guaranteed by the JulianDate constructors, double checked here.
+  assert(dayNumber      >= 0,             s"dayNumber >= 0")
+  assert(nanoAdjustment >= MinAdjustment, s"nanoAdjustment >= $MinAdjustment")
+  assert(nanoAdjustment <= MaxAdjustment, s"nanoAdjustment <= $MaxAdjustment")
+
+
+  /** Julian date value as a Double, including Julian Day Number and fractional
+    * day since the preceding noon.
+    */
+  val toDouble: Double =
+    dayNumber + nanoAdjustment.toDouble / NanoPerDay.toDouble
+}
+
+
+object JulianDate {
+
+  // One half day of seconds, ignoring leap seconds since Java time API ignores
+  // them as well.
+  private val SecPerDay: Int      = 86400
+  private val SecPerHalfDay: Int  = 43200
+
+  private val Billion: Int        = 1000000000
+  private val NanoPerDay: Long    = SecPerDay.toLong * Billion.toLong
+
+  private val MinAdjustment: Long = -SecPerHalfDay.toLong * Billion.toLong
+  private val MaxAdjustment: Long =  SecPerHalfDay.toLong * Billion.toLong - 1
+
+  /** Convert an `Instant` to a Julian Date.
+    */
+  def ofInstant(i: Instant): JulianDate =
+    ofLocalDateTime(LocalDateTime.ofInstant(i, UTC))
+
+  /** JulianDate from a `LocalDateTime` assumed to represent a time at UTC.
+    */
+  def ofLocalDateTime(ldt: LocalDateTime): JulianDate = {
+    val y   = ldt.getYear
```

Comment:
Where are these formulas? on the wikipedia page?

Reply:
The computation for Julian Day is reproduced in the wikiwiki page and elsewhere.  They attribute it to an "explanatory supplement" of an astronomical almanac from 1992 where it is attributed to "Fliegel and Van Flandern (1968)":

![screenshot 2018-01-04 09 42 07](https://user-images.githubusercontent.com/4906023/34564235-a6105d8c-f134-11e7-8ab5-5099c474975e.jpeg)

I suppose to be correct it should check that the date is actually actually after -4713 November 23 and return the result wrapped in a corresponding `Option`.  Or maybe that is overkill.

The fractional part is just the fraction of a day corresponding to the time.

Reply:
Fliegel, H.F. and Van Flandern, T.C. (1968). "A Machine Algorithm for Processing Cal-
endar Dates" Communications ol the Association of Computing Machines ll, 6sT.